File: 19. Implement Min Stack with space optimization.cpp
/*
 Task: Implement Min Stack with space optimization
*/

#include <bits/stdc++.h>
using namespace std;

// Store differences to reduce extra stack space:
// https://stackoverflow.com/questions/1603095/how-to-implement-a-stack-with-getmin
class MinStackOpt
{
   public:
    void push(long long x)
    {
        if (st.empty())
        {
            st.push(x);
            minVal = x;
        }
        else
        {
            if (x < minVal)
            {
                st.push(2 * x - minVal);
                minVal = x;
            }
            else
                st.push(x);
        }
    }
    bool pop(long long& out)
    {
        if (st.empty())
            return false;
        long long t = st.top();
        st.pop();
        if (t < minVal)
        {
            out = minVal;
            minVal = 2 * minVal - t;
        }
        else
            out = t;
        return true;
    }
    long long top()
    {
        if (st.empty())
            throw runtime_error("empty");
        long long t = st.top();
        return (t < minVal) ? minVal : t;
    }
    long long getMin()
    {
        if (st.empty())
            throw runtime_error("empty");
        return minVal;
    }
    bool empty() const { return st.empty(); }

   private:
    stack<long long> st;
    long long minVal;
};

int main()
{
    MinStackOpt ms;
    ms.push(3);
    ms.push(2);
    ms.push(1);
    cout << ms.getMin() << "\n";
    long long v;
    ms.pop(v);
    cout << ms.getMin() << "\n";
    return 0;
}

// Time Complexity: push/pop/top/getMin O(1)
// Space Complexity: O(n) but optimized

File: 19. Implement Min Stack with space optimization.py
"""
Task: Implement Min Stack with space optimization

Time Complexity: push/pop/top/getMin O(1)
Space Complexity: O(n) but optimized
"""

class MinStackOpt:
    def __init__(self):
        self.st = []
        self.min_val = None

    def push(self, x: int):
        if not self.st:
            self.st.append(x)
            self.min_val = x
        else:
            if x < self.min_val:
                # store coded value
                self.st.append(2*x - self.min_val)
                self.min_val = x
            else:
                self.st.append(x)

    def pop(self):
        if not self.st:
            raise IndexError('pop from empty stack')
        t = self.st.pop()
        if t < self.min_val:
            orig = self.min_val
            self.min_val = 2*self.min_val - t
            return orig
        else:
            return t

    def top(self):
        if not self.st:
            raise IndexError('top from empty stack')
        t = self.st[-1]
        return self.min_val if t < self.min_val else t

    def get_min(self):
        if not self.st:
            raise IndexError('min from empty stack')
        return self.min_val

if __name__ == '__main__':
    ms = MinStackOpt()
    ms.push(3)
    ms.push(2)
    ms.push(1)
    print(ms.get_min())
    ms.pop()
    print(ms.get_min())
